src/fabric.c

```diff
@@ -512,6 +512,86 @@ int ofi_nic_close(struct fid *fid)
 	return 0;
 }
 
+struct fi_ops default_nic_ops = {
+	.size = sizeof(struct fi_ops),
+	.close = ofi_nic_close,
+};
+
+static int ofi_dup_dev_attr(const struct fi_device_attr *attr,
+			    struct fi_device_attr *dup_attr)
+{
+	if (str_dup(attr->name, &dup_attr->name) ||
+	    str_dup(attr->device_id, &dup_attr->device_id) ||
+	    str_dup(attr->device_version, &dup_attr->device_version) ||
+	    str_dup(attr->vendor_id, &dup_attr->vendor_id) ||
+	    str_dup(attr->driver, &dup_attr->driver) ||
+	    str_dup(attr->firmware, &dup_attr->firmware))
+		return -FI_ENOMEM;
+
+	return 0;
+}
+
+static int ofi_dup_link_attr(const struct fi_link_attr *attr,
+			     struct fi_link_attr *dup_attr)
+{
+	if (str_dup(attr->address, &dup_attr->address) ||
+	    str_dup(attr->network_type, &dup_attr->network_type))
+		return -FI_ENOMEM;
+
+	dup_attr->mtu = attr->mtu;
+	dup_attr->speed = attr->speed;
+	dup_attr->state = attr->state;
+	return 0;
+}
+
+struct fid_nic *ofi_nic_dup(const struct fid_nic *nic)
+{
+	struct fid_nic *dup_nic;
+	int ret;
+
+	dup_nic = calloc(1, sizeof(*dup_nic));
+	if (!dup_nic)
+		return NULL;
+
+	dup_nic->device_attr = calloc(1, sizeof(*dup_nic->device_attr));
+	dup_nic->bus_attr = calloc(1, sizeof(*dup_nic->bus_attr));
+	dup_nic->link_attr = calloc(1, sizeof(*dup_nic->link_attr));
```

Comment:
I will update to make the function follow the same semantics that we have for dupinfo.
The fields are pointers, so that each attribute structure can be extended in the future separately without affecting the offsets of fields in unrelated structures.  E.g. we can add a new bus type with new attributes without breaking the link_attr access.

Reply:
But doesn't that just lead to *more* versioning?  I.e., you have to version each of those individual structs (*and* the top-level struct) vs. just versioning the top-level struct.

Reply:
We use a top level API version (e.g. from fi_getinfo) to know if we can access a field or not in any input structure (cq_attr, ep_attr, tx_attr, etc.).  For output structures, we can add new fields to the end without version checks.

In this case, a fid_nic is an output structure from the provider.  It's not something an app can craft and expect to pass down.  So, we can add new fields to device_attr without impacting how an application will access the bus_attr or link_attr fields.  If we fold everything into a single structure, the provider would actually need to deal with fid_nic_v1 and fid_nic_v2 etc.  The proposal allows them to always just use the latest structure definition independent of what version the app understands.